cmd/keymasterd/awsRole.go

```diff
@@ -0,0 +1,341 @@
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/Cloud-Foundations/keymaster/lib/certgen"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/organizations"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+const (
+	awsAccountListInterval = time.Minute * 5
+)
+
+type assumeRoleCredentialsProvider struct {
+	credentials aws.Credentials
+	roleArn     *string
+	stsClient   *sts.Client
+}
+
+type getCallerIdentityResult struct {
+	Arn string
+}
+
+type getCallerIdentityResponse struct {
+	GetCallerIdentityResult getCallerIdentityResult
+}
+
+type parsedArnType struct {
+	parsedArn arn.ARN
+	role      string
+}
+
+func awsListAccounts(ctx context.Context, orgClient *organizations.Client) (
+	map[string]struct{}, error) {
+	output, err := orgClient.ListAccounts(ctx,
+		&organizations.ListAccountsInput{})
+	if err != nil {
+		return nil, err
+	}
+	list := make(map[string]struct{}, len(output.Accounts))
+	for _, account := range output.Accounts {
+		list[*account.Id] = struct{}{}
+	}
+	return list, nil
+}
+
+func getCallerIdentity(header http.Header,
+	validator func(presignedUrl string) error) (*parsedArnType, error) {
+	claimedArn := header.Get("claimed-arn")
+	presignedMethod := header.Get("presigned-method")
+	presignedUrl := header.Get("presigned-url")
+	if claimedArn == "" || presignedUrl == "" || presignedMethod == "" {
+		return nil, fmt.Errorf("missing presigned request data")
+	}
+	if err := validator(presignedUrl); err != nil {
+		return nil, err
+	}
+	validateReq, err := http.NewRequest(presignedMethod, presignedUrl, nil)
+	if err != nil {
+		return nil, err
+	}
+	validateResp, err := http.DefaultClient.Do(validateReq)
+	if err != nil {
+		return nil, err
+	}
+	defer validateResp.Body.Close()
+	if validateResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("verification request failed")
+	}
+	body, err := ioutil.ReadAll(validateResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var callerIdentity getCallerIdentityResponse
+	if err := xml.Unmarshal(body, &callerIdentity); err != nil {
+		return nil, err
+	}
+	parsedArn, err := arn.Parse(callerIdentity.GetCallerIdentityResult.Arn)
+	if err != nil {
+		return nil, err
+	}
+	parsedArn.Region = ""
+	parsedArn.Service = "iam"
+	splitResource := strings.Split(parsedArn.Resource, "/")
+	if len(splitResource) < 2 || splitResource[0] != "assumed-role" {
+		return nil, fmt.Errorf("invalid resource: %s", parsedArn.Resource)
+	}
+	parsedArn.Resource = "role/" + splitResource[1]
+	if parsedArn.String() != claimedArn {
+		return nil, fmt.Errorf("validated ARN: %s != claimed ARN: %s",
+			parsedArn.String(), claimedArn)
+	}
+	return &parsedArnType{
+		parsedArn: parsedArn,
+		role:      splitResource[1],
+	}, nil
+}
+
+func validateStsPresignedUrl(presignedUrl string) error {
+	parsedPresignedUrl, err := url.Parse(presignedUrl)
```

Comment:
to calm down the validator... can you return the parsed url and rebuild it on the caller?

Reply:
Done. Hopefully this shuts it up.

Reply:
Wow. It worked. It shut up the code scanner.